### DISCUSSION TITLE
Redraw/Invalidate after Diagram Container Hide/Show

### DISCUSSION THREAD

**Discussion by cerulean-tiger** [clientIO/joint#2108] [Q&A] [answered]
### Introduction

I am developing a page with more than one diagram.  I happen to be using Bootstrap tabs/pills to toggle views of which diagram is shown.  I'm seeing that when a diagram container has css "display: none" and then later shown that the links in the diagram are not connected to their nodes/elements at the correct locations.  Even if you are not using Bootstrap and manually alter the display attribute of the css for the container or one of its parent elements, the diagram does not render correctly when shown.  My question is if there is a paper or graph function that can be invoked to invalidate the current view and redraw all the links and nodes anew?

### Steps to reproduce

1. Create a div for your first diagram in the page
2. Create a div for your second diagram in the page, but add css styling "display: none" to it
3. Run code to initialize both diagrams on the page including links from/to ports on node elements.
4. Dynamically change the css so that the first diagram is now "display: none" and the second diagram div no longer has this css property.
5. Observe that the link locations (source/target) may not be connected to the ports defined on the nodes.

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS

**Comment by kumilingus** [comment, marked as answer]
You need to call `paper.freeze()` before hiding the paper and `paper.unfreeze()` after you show it.

Here's the **bootstrap tabs** demo: https://www.jointjs.com/demos/bootstrap-tabs

A related issue: https://github.com/clientIO/joint/issues/407
